Skip non-delimited data at the top of amplitude data files

jonocarroll opened the pull request:
I'm a bit new to this data, but for a while it seemed Bio-Rad dropped support for exporting the full set of data, instead only offering the `.ddpcr` file (a password-protected 7zip file, presumably containing the `.csv` data). They've rectified that in a much more recent version, but unless there's a different setting for it, the data now has some non-delimited header rows, e.g.
```
Target Value of 0 = negative
Target Value of 1 = positive
Target Value of u = unclassified (Advanced Classification Mode)

Ch1Amplitude,Ch2Amplitude,1,2,
9265.351,781.6206,1,0,
```
The column names are also different to what {ddpcr} is expecting, but one problem at a time.

This PR adds some checking of the data to see if there's anything that needs to be skipped,  and if so, skips it. Thanks to the trailing comma, {readr} thinks there's a blank column which I've dropped. I also added some silencing of the {readr} messages.

No impact to reading the older data (no changes to test outputs) and I'm able to at least generate a plate object from the newer data.

daattali commented on R/utils.R:
```diff
@@ -412,9 +412,34 @@ capitalize <- function(x) {
 }
 
 # At some point around 2019, read_csv introduced new attributes that broke tests
+# NOTE: the first argument should be the path to the file to be read
```
Thanks for this note!

daattali commented on R/utils.R:
```diff
@@ -412,9 +412,34 @@ capitalize <- function(x) {
+# Target Value of 0 = negative
+# Target Value of 1 = positive
+# Target Value of u = unclassified (Advanced Classification Mode)
+#  -- this detects that they should be skipped and otherwise
```
"this detects that they should be skipped" -> "this detects and returns the number of lines that should be skipped"

daattali commented on R/utils.R:
```diff
@@ -412,9 +412,34 @@ capitalize <- function(x) {
+detect_skip <- function(file) {
+  n <- 0
+  found_data <- FALSE
+  while (!found_data) {
```
Does this always return, even if it never finds a correct line? Would it loop infinitely if a bad file is given?

jonocarroll replied:
It loops through until it finds a non-blank line that doesn't start with "Target", which should be the column header or the data itself. The only case I can think of that would infinitely loop would be a file consisting of only lines starting with "Target", but I can add some safety nonetheless.

daattali reviewed (changes requested):
Can you add tests for a file that uses this new format, and also for a file that may never find the Target column?

jonocarroll commented on R/utils.R:
```diff
@@ -412,9 +412,34 @@ capitalize <- function(x) {
+  # silence readr repairing names
+  data <- readr::read_csv(..., skip = skips, name_repair = "unique_quiet", show_col_types = FALSE)
+  # drop empty columns
+  data <- data[, vapply(data, function(x) !all(is.na(x)), TRUE)]
```
This is because the new format has trailing commas in the data so `readr` thinks there's another column there.
```
Ch1Amplitude,Ch2Amplitude,1,2,
9265.351,781.6206,1,0,
```

jonocarroll commented:
I've added some test data from Bio-Rad with the new format. I truncated the files down to the first 100 lines for the sake of file size.

Note that this reads in as 32 wells not 16, presumably because the metadata file lists each well twice. If I change the filename of the metadata file so it isn't automatically detected, the plate object shows 16 wells.

daattali reviewed (changes requested):
The test files need a little more work to make them more unittest-type rather than so tightly coupled with a particular dataset. This means:
- A more meaningful folder name that describes what it's testing instead of "SMN2CNV" which offers no information
- The very long file files have a lot of irrelevant information (I understand they were probably taken from a real run, but this should be stripped away when put into a test suite)
- No need for 16 different wells, just a few would suffice
- The meta file name should also be more descriptive, rather than using the string "752.csv" as its identifier
- A few test cases should be added for files where data is never found (all lines are blank or start with "Target"

jonocarroll commented on R/utils.R:
```diff
@@ -412,9 +412,35 @@ capitalize <- function(x) {
+# Target Value of 1 = positive
+# Target Value of u = unclassified (Advanced Classification Mode)
+#  -- this detects and returns the number of lines that should be skipped
+detect_skip <- function(file) {
+  n <- 0
+  found_data <- FALSE
+  lines_in_file <- length(readLines(file, warn = FALSE))
+  while (!found_data && n <= lines_in_file) {
+    n <- n + 1
+    d <- readLines(file, n = n)
+    if (!grepl("^Target|^$", utils::tail(d, 1))) {
+      found_data <- TRUE
+    }
+  }
+  if (n == lines_in_file) stop("Reached end of file without detecting any data")
+  n - 1
+}
```
That's a good point, and a potential win for the tests I should have written. I'll update.

jonocarroll commented:
I've finally come back to this, sorry. I added a test for empty data and trimmed down the example data.

daattali commented:
Can yo bring this PR up to date with master? The first change I see in this PR right now is reverting a line I added to the `DESCRIPTION` file, which might mean that there are more changes being overwritten

daattali commented:
Don't worry about my previous comment, I made a couple small changes, just make sure to pull before you do any more work.

I still think the tests could use some improvement.

- Is "cnv" a descriptive name to what this feature is testing? If this new data format only happens for CNV datasets, then this name does make sense. But if all new exports have these "Target ..." lines, then the name of this dataset should reflect that, so that we know what this is testing. For example, there's a test folder "read_v174" - that was added when testing a feature that was added in quantasoft version 1.7.4. It's not super descriptive but at least years later I can look at this test and have a clue about what it's for
- The test files should be minimized and simplified. A few lines in each test file is enough, there's no benefit in having 100 lines, if anything it just makes debugging harder if a test fails
- The tests for `detect_skip()` should be moved into the `test-utils.R` file 
- The tests for `detect_skip()` should be more comprehensive. If I were to test that function, I would test with files like these to make sure it works as expected:

```



```

```
Target 1
Target 2
```

```
data
Target 1
Target 2
```

```
Target1
data
Target2
```

```
Target1
Target2
data
```

```
Target1


Target2
data
```

```


Target1
Target2
data
```

daattali commented:
Any plans to revive/complete this?

jonocarroll commented:
I'll see what I can do to bring it up to spec - the use case has dropped off my radar but I'd like to wrap up what I can.

daattali commented:
Do you still want to try to finish this PR?

jonocarroll commented:
I _am_ sorry that I dropped the ball on this - I'll try to have an update for you by the end of next week, but I'm not actively using this file format any longer so I can't comment on how common it is to have these changes. If it's helpful, I can move these changes to an independent function so it can be developed without interfering with the original.

daattali commented:
If you're not using it, and noone else complained about it, and the current version restored support - then I'm not sure it's worth your time to work on it. Feel free to do continue the PR if you're up for it, but I'm also ok without it.

jonocarroll commented:
> Is "cnv" a descriptive name to what this feature is testing? 

I changed the shortname to `quant` because I think this header is identifying the automated quantification. I don't have an insight into how this header came about or how frequent it is, but this PR makes this package robust to that being there (or not) without intervention.

> The test files should be minimized and simplified.

✅  Done.

> The tests for detect_skip() should be moved into the test-utils.R file

✅  Done - I had to copy the `testdir()` helper there as well.

> The tests for detect_skip() should be more comprehensive

I believe the header is all-or-nothing; it's either there with those three lines and a space, or the file starts with the data. The tests currently cover those two cases.

---

All tests currently pass.

daattali commented:
Thanks for finishing this up!
